### PR TITLE
[6X] Preserve gp_fastsequence during upgrade

### PR DESCRIPTION
**Pull request by kyeap-vmware**
The gp_fastsequence catalog table must be preserved when upgrading to a newer version. Not preserving this table will likely cause duplicate ctids in ao tables when normal cluster operations resume after an upgrade.

How duplicate ctids can occur:
1. gp_fastsequence is initialized to 0 for a new ao table.
2. gp_fastsequence only increases positively in increments of 100.
3. ctids are assigned based on gp_fastsequence values.

When a tuple is inserted it consumes a gp_fastsequence value to produce a ctid to index the data. If a fastsequence value is reused, a duplicate ctid will occur. This will result in indexes returning wrong results.

When an ao table's fastsequence value is lower than the it's tuple count, gpdb's default behavior in an attempt to fix the fastsequence value is to set it to the table's tuple count on the next insert. However this not reliable since fastsequence reserves numbers in increments of 100 at a time for performance purposes. This means that even if only 1 value is inserted into the table, fastsequence will increment by 100. As a result Under normal cluster operation, fastsequence increments at a much faster rate than tuple count and setting it to a table's tuple count can still result in a duplicate ctid because there will be rows with ctids based on fastsequence values much higher than its table's tuple count.

The alternative to relfilenode approach is to do something similar to restore_aosegment_tables() where we would scan all gp_fastsequence in all databases, generate UPDATE statements to manually rebuild the catalog table.

Test is located in gpupgrade PR: https://github.com/greenplum-db/gpupgrade/pull/736

gpupgrade: https://cm.ci.gpdb.pivotal.io/teams/main/pipelines/gpupgrade:fastsequence-test
6X: https://cm.ci.gpdb.pivotal.io/teams/main/pipelines/upgrade-gp-fastsequence

### CODE REVIEW

File: contrib/pg_upgrade/greenplum/frozenxids_gp.c [hunk 1]
```diff
@@ -17,21 +17,21 @@
  * of the relation in pg_class with the datafrozenxid from the corresponding
  * database in the old cluster. This ensures that the xmin is not > relfrozenxid
  * for any of the tuple.
- * Also, update the new segment database with the datafrozenxid
+ * Coordinator contains data entries in gpdb catalog tables like gp_fastsequence
+ * and pg_ao(cs)seg, and pg_aovisimap that also need relfrozenxid correction.
+ * Also, update the new database with the datafrozenxid
```
Comment (jimmyyih) [inline contrib/pg_upgrade/greenplum/frozenxids_gp.c:22]: s/datafrozenxid/datfrozenxid/ (two instances of this L17 and L22)

The wording can be a bit more clear too:
```
 * Coordinator contains data entries in GPDB catalog tables like gp_fastsequence,
 * pg_aoseg, pg_aocsseg, and pg_aovisimap that also need relfrozenxid correction.
 * Also, update the new database with the datfrozenxid
```

**Review by jimmyyih — APPROVED** [review]

**Comment by ashwinstar** [comment]
The description directly jumps to this statement after explaining the problem "The alternative to relfilenode approach is to do something similar to restore_aosegment_tables() where we would scan all gp_fastsequence in all databases, generate UPDATE statements to manually rebuild the catalog table." seems missing to describe the current "relfilenode approach" in middle?

File: contrib/pg_upgrade/info.c [hunk 2]
```diff
@@ -519,7 +519,8 @@ get_rel_infos(ClusterInfo *cluster, DbInfo *dbinfo)
 			 "    n.nspname NOT IN ('gp_toolkit', 'pg_bitmapindex') AND "
 			 "	  c.oid >= %u) "
 			 "  OR (n.nspname = 'pg_catalog' AND "
-	"    relname IN ('pg_largeobject', 'pg_largeobject_loid_pn_index'%s) ));",
+	"    relname IN ('pg_largeobject', 'pg_largeobject_loid_pn_index'%s, "
+	"                'gp_fastsequence', 'gp_fastsequence_objid_objmod_index') ));",
```
Comment (ashwinstar) [inline contrib/pg_upgrade/info.c:523]: seems like the PR should be split into 2 commits
1 fixing the regular coordinator specific issues which while working on fast sequence were revealed and are being fixed
2. actual work needed to preserve gp_fastsequence which seems just this line

having them separate I think would be good for longer term - lets say for whatever reason we decide to revert preserving fast sequence still the other changes made in this PR for coordinator freezing and such are important and necessary fixes to keep, is that assessment correct?

File: contrib/pg_upgrade/greenplum/frozenxids_gp.c [hunk 3]
```diff
@@ -14,24 +14,24 @@
  * Segment database contains data, and the tuples should not have
  * any entry with a xmin > relfrozenxid for the table in pg_class.
  * Instead of vacuum freezing the entire data, update the relfrozenxid
- * of the relation in pg_class with the datafrozenxid from the corresponding
+ * of the relation in pg_class with the datfrozenxid from the corresponding
  * database in the old cluster. This ensures that the xmin is not > relfrozenxid
  * for any of the tuple.
- * Also, update the new segment database with the datafrozenxid
+ * Coordinator contains data entries in GPDB catalog tables like gp_fastsequence,
```
Comment (ashwinstar) [inline contrib/pg_upgrade/greenplum/frozenxids_gp.c:20]: Surprising coordinator contains entries for gp_fastsequence, seems wasteful. Need to look separately to at least improve this aspect for 7X and forward.

**Review by kalensk — APPROVED** [review]
LGTM.

**Review by ashwinstar — APPROVED** [review]

**Pull request merged**